pallets/roles/src/lib.rs

```diff
@@ -0,0 +1,239 @@
+
+const ROLES_STAKING_ID: LockIdentifier = *b"rstaking";
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	/// Configuration trait.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		type Currency: LockableCurrency<
+			Self::AccountId,
+			Moment = BlockNumberFor<Self>,
+			Balance = Self::CurrencyBalance,
+		>;
+		/// Just the `Currency::Balance` type; we have this item to allow us to constrain it to
+		/// `From<u64>`.
+		type CurrencyBalance: sp_runtime::traits::AtLeast32BitUnsigned
+			+ codec::FullCodec
+			+ Copy
+			+ MaybeSerializeDeserialize
+			+ sp_std::fmt::Debug
+			+ Default
+			+ From<u64>
+			+ TypeInfo
+			+ MaxEncodedLen;
+
+		/// Handler for the unbalanced reduction when slashing a staker.
+		type Slash: OnUnbalanced<NegativeImbalanceOf<Self>>;
+
+		type WeightInfo: WeightInfo;
+
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Role assigned to the validator.
+		RoleAssigned { account: T::AccountId, role: RoleType },
+		/// Removed validator from role.
+		RoleRemoved { account: T::AccountId, role: RoleType },
+		/// Funds bonded to become a validator.
+		Bonded { account: T::AccountId, amount: BalanceOf<T> },
+		/// Funds unbonded to stop being a validator.
+		Unbonded { account: T::AccountId, amount: BalanceOf<T> },
+		/// Slashed validator.
+		Slashed { account: T::AccountId, amount: BalanceOf<T> },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Role has already been assigned to provided validator.
+		RoleAlreadyAssigned,
+		/// No role assigned to provided validator.
+		RoleNotAssigned,
+		/// Insufficient bond to become a validator.
+		InsufficientBond,
+		/// Stash controller account already added to Ledger
+		AlreadyPaired,
+		/// Stash controller account not found in ledger
+		InvalidStashController,
+	}
+
+	/// Map from all "controller" accounts to the info regarding the staking.
+	#[pallet::storage]
+	#[pallet::getter(fn ledger)]
+	pub type Ledger<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AccountId, RoleStakingLedger<T>>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn account_role)]
+	/// Mapping of resource to bridge index
+	pub type AccountRolesMapping<T: Config> = StorageMap<_, Blake2_256, T::AccountId, RoleType>;
+
+	/// The minimum active bond to become and maintain the role.
+	#[pallet::storage]
+	#[pallet::getter(fn min_active_bond)]
+	pub(super) type MinActiveBond<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight({0})]
+		#[pallet::call_index(0)]
+		pub fn assign_role(
```

Comment:
Ensure that only validators can get roles. Possibly a small extra bond is useful, but most importantly is that I can restake my existing stake. This means, I subject my validator stake to these new slashing conditions.